### ISSUE TITLE
Returning not captured by tracker

### ISSUE THREAD

**Issue by mir1198yusuf** [felixmosh/knex-mock-client#20]
I have an insert query with `returning('*')`.
Insert is captured in `tracker.on.any` but not `returning`.
Any suggestions ?

**Comment by felixmosh** [comment]
Can you share the code & the test?

**Comment by mir1198yusuf** [comment]
`
await ModelName.insert(params)
.returning('*')
.first()
`

**Comment by mir1198yusuf** [comment]
Another case
I tested a query involving select & graph fetch, only select was captured.
`
await ModelName.query()
.select('abc')
.withGraphFetch('relationName')
`

Only select was captured.
Select of graph fetch is not captured.

**Comment by mir1198yusuf** [comment]
I am using objection-js so it internally uses knex.

How should I do it ?

**Comment by felixmosh** [comment]
I'm not familiar with `pg` so much, but let me investigate it.
Did you passed the `dialect` property?

**Comment by felixmosh** [comment]
I've tried to run insert with returning,
```js
await db('table_name')
        .insert({ id: 1234, created_at: '2022-05-10', updated_at: '2022-05-10' })
        .returning('*');
```
Printed out the call history, and I do see insert catch with the returning part.
```js
{
  method: 'insert',
  options: {},
  timeout: false,
  cancelOnTimeout: false,
  bindings: [Array],
  __knexQueryUid: 'BEJ_r5crW6qLctx8BmrVY',
  sql: 'insert into "table_name" ("created_at", "id", "updated_at") values (?, ?, ?) returning *',
  returning: '*',
  queryContext: undefined,
}
```

**Comment by mir1198yusuf** [comment]
This `returning` you have used is `objection-js` library's ?

**Comment by mir1198yusuf** [comment]
Just read this https://github.com/felixmosh/knex-mock-client/issues/20#issuecomment-1128503058.
Objection generates a new query for graph fetch.
Just to give the working flow, in `mock-knex` graph fetch queries are captured/tracked as separate query (different `step` number).

I don't know but for some reason, graph fetch queries (basically extra select queries) are not captured by this mock client.

**Comment by felixmosh** [comment]
Sounds like you are not returning the data that objection requires for its graph fetch... but it is just a guess.
Can you prepare a small repo with Objection & knex-mock-client?

**Comment by mir1198yusuf** [comment]
Perfect guess!
I also found that I was not returning the join field value that is needed for graph fetch query.
When I added that, it worked!
Thanks.

**Comment by felixmosh** [comment]
Can we close this issue?

**Comment by mir1198yusuf** [comment]
Also, the main issue was for `returning`.

How I used it was `await ModelName.query().insert(params) .returning('*') .first()`. This is in objection.
Above worked fine. I can now see `returning` in generated sql.

But if I add transaction to objection query, `insert` is captured but `returning` is missing in generated sql.
See below snip.

`const knexRead  = await knex({
         //.....config values
})`

```
const trx = await knexRead.transaction()
`await ModelName.query(trx).insert(params) .returning('*') .first()`
```

In actual code (real interaction with db), returning works & returns data but in unit-test mocking, returning does not appear in generated sql.

**Comment by felixmosh** [comment]
There are test with [transactions](https://github.com/felixmosh/knex-mock-client/blob/fa719ed55cdcb23f60cb449d2aaf3d4f4ac5f8b8/tests/insert.spec.ts#L195), which works with nested syntax
```js
await db.transaction(async (trx) => {
  await db('table_name').insert({ name: faker.name.firstName() }).transacting(trx);
  await db('table_name').delete().where({ name: faker.name.firstName() }).transacting(trx);
});
```

Maybe there is an issue with the commit style, I'll check it.

**Comment by mir1198yusuf** [comment]
Okay sure! 
Just let me know that.
If it does not work commit style, I may do unit-test without transaction.

**Comment by felixmosh** [comment]
I've extracted all transaction tests to a separate spec, and added one with `commit` style.
https://github.com/felixmosh/knex-mock-client/blob/master/tests/transaction.spec.ts
It works as expected. (actually, the mock ignores transactions by [empty implementation](https://github.com/felixmosh/knex-mock-client/blob/master/src/Tracker.ts#L43-L48))

**Comment by mir1198yusuf** [comment]
> The returning part is related to `dialect` property, can you check that you are passing it?

I am passing `pg` dialect.

**Comment by mir1198yusuf** [comment]
No problem! 
Meanwhile I would unit-tests without transaction.

**Comment by felixmosh** [comment]
fix in `v1.8.2`

**Comment by mir1198yusuf** [comment]
Tested! Working good.

**Comment by mir1198yusuf** [comment]
I am getting this message, not an issue though.
`(node:62616) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 start listeners added to [MockClient]. Use emitter.setMaxListeners() to increase limit`

If this is of any concern, I can create a new issue. If not let it be. 
Is this due to multiple `async` unit-tests service function running in parallel by jest ?

**Comment by felixmosh** [comment]
Can you open a new issue with this, I'll check it.

**Comment by colvint** [comment]
> I am getting this message, not an issue though. `(node:62616) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 start listeners added to [MockClient]. Use emitter.setMaxListeners() to increase limit`
> 
> If this is of any concern, I can create a new issue. If not let it be. Is this due to multiple `async` unit-tests service function running in parallel by jest ?

`MockClient.EventEmitter.defaultMaxListeners = 100`

**Comment by felixmosh** [comment]
@colvint can you open a new issue with details of how to reproduce this error...

**Comment by colvint** [comment]
It's rather involved to repro. Given the warning, I presume if you add enough trackers in _just_ enough `beforeEach`s, you'll trigger it. Above I'm pointing out one can silence the warnings by upping the number of event emitter listeners.
